src/components/implementation/sched/fprr/sched.c

```diff
@@ -0,0 +1,98 @@
+#include <cos_defkernel_api.h>
+#include <sched.h>
+#include <sl.h>
+#include "sched_info.h"
+
+int
+sched_thd_wakeup_intern(spdid_t c, thdid_t t)
+{
+	if (!c || !sched_childinfo_find(c)) return -1;
+	sl_thd_wakeup(t);
+
+	return 0;
+}
+
+int
+sched_thd_block_intern(spdid_t c, thdid_t deptid)
+{
+	if (!c || !sched_childinfo_find(c)) return -1;
+	sl_thd_block(deptid);
+
+	return 0;
+}
+
+int
+sched_thd_block_timeout_intern(spdid_t c, thdid_t deptid, u32_t hi, u32_t lo)
+{
+	if (!c || !sched_childinfo_find(c)) return -1;
+	/* TODO: return time elapsed */
+	sl_thd_block_timeout(deptid, ((cycles_t)hi << 32 | (cycles_t)lo));
+
+	return 0;
+}
+
+thdid_t
+sched_thd_create_intern(spdid_t c, int idx)
+{
+	struct cos_defcompinfo *dci;
+	struct sl_thd *t = NULL;
+
+	if (!c) return 0;
+	dci = sched_child_defci_get(sched_childinfo_find(c));
+	if (!dci) return 0;
```

Comment:
That is something I've really not thought about: deletion of a component. There is no interface for doing such a thing yet. I'll make a `/* FIXME: need looks when we allow child components to be deallocated */` sort of if that's ok. 
I feel like having locks here is unnecessary otherwise.

Reply:
I'm more concerned about if you've thought through the concurrency issues.  Again, I'd really rather you didn't design fundamental components that are going to be used for a long time with a very narrow range of capabilities.  I don't want someone to modify these and have un-traceable bugs because you didn't put two lines of code in there now.  Design should not only consider the current code, but future uses of the code.